src/adaptation/stepsize.jl

```diff
@@ -0,0 +1,117 @@
+######################
+### Mutable states ###
+######################
+
+mutable struct DAState{T<:AbstractFloat}
+    m     :: Int
+    ϵ     :: T
+    μ     :: T
+    x_bar :: T
+    H_bar :: T
+end
+
+function DAState(ϵ::AbstractFloat)
+    μ = computeμ(ϵ)
+    return DAState(0, ϵ, μ, 0.0, 0.0)
+end
+
+function computeμ(ϵ::AbstractFloat)
+    return log(10 * ϵ) # see NUTS paper sec 3.2.1
+end
+
+function reset!(dastate::DAState{T}) where {T<:AbstractFloat}
+    dastate.μ = computeμ(dastate.ϵ)
+    dastate.m = 0
+    dastate.x_bar = zero(T)
+    dastate.H_bar = zero(T)
+end
+
+mutable struct MSSState{T<:AbstractFloat}
+    ϵ :: T
+end
+
+################
+### Adapters ###
+################
+
+abstract type StepSizeAdapter <: AbstractAdapter end
+
+struct FixedStepSize{T<:AbstractFloat} <: StepSizeAdapter
+    ϵ :: T
+end
+
+function getss(fss::FixedStepSize)
+    return fss.ϵ
+end
+
+struct DualAveraging{T<:AbstractFloat} <: StepSizeAdapter
```

Comment:
can we make `DualAveraging` an immutable type, and let 

- `adapt_stepsize!`, `adapt!`
- `finalize!` 

always return a new `DualAveraging` instance. This can improve both readability and performance since immutable types are allocated on stacks. If so we can also remove

- `reset!` 

and merge 

- `adapt!(da::DualAveraging, θ::AbstractVector{<:AbstractFloat}, α::AbstractFloat)`
- `adapt_stepsize!(da::DualAveraging, α::AbstractFloat)` 

into 

-`adapt(d::AbstractAdaptor, α::AbstractFloat, θ::AbstractVector{<:AbstractFloat}=zeros(0))`

We can probably do similar things for `Metric` adaptors.